samples: mesh: nrf52: Simplified implementation

vikrant8052 opened the pull request:
Create single structure of light_ctl_state & simplified state binding algorithm.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: Simplified implementation

Create single structure of light_ctl_state & simplified
state binding algorithm.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

LingoMeng commented:
@vikrant8051 
If each model's dependencies are defined in `Kconfig`, that would be best. For example: I only want to associate the` Light CTL Server model` with `Kconfig `to automatically associate other models that it depends on. I think this way is the most friendly way to look forward to your point of view.

vikrant8052 committed:
sample: mesh: nrf52: complete inst. transition without timer

If transition time (tt) estimated equal to zero then transition
get completed without using timer.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@trond-snekvik @jhedberg 
Please review this PR.

trond-snekvik reviewed (changes requested):
Sorry for the slow response time on this one. 

There are a lot of changes here, and I'm not confident that I've gone through them well enough to catch everything. I found one proper bug that needs fixing, but the rest of my comments are just nitpicks.

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -324,29 +242,28 @@ static void gen_onoff_status(struct bt_mesh_model *model,
-			net_buf_simple_add_le16(msg, state->target_level);
+		if (state->transition->type == LEVEL_TT_MOVE) {
+			net_buf_simple_add_le16(msg, (s16_t) get_target(LEVEL));
 			net_buf_simple_add_u8(msg, UNKNOWN_VALUE);
```
set_target() is already taking care of lightness range here.

trond-snekvik replied:
Yes, but shouldn't the transition time have a known value?

vikrant8051 replied:
Transition will not end as it reaches to end of range. That's why in case of gen_move rt = Unknown.

vikrant8051 replied:
using commit no. 5 I have corrected Gen. move implementation.

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -179,90 +170,18 @@ static void gen_onoff_set_unack(struct bt_mesh_model *model,
-		} else {
-			gen_onoff_publish(model);
-			return;
+		if (model->op->opcode == BT_MESH_MODEL_OP_GEN_ONOFF_SET) {
```
This App is not allowing me to provision it. That's why not able to test end result.
I've already raised issue here: https://github.com/zephyrproject-rtos/zephyr/issues/19685

vikrant8051 replied:
```
/* Message handler for the opcode */
	void (*const func)(struct bt_mesh_model *model,
			   struct bt_mesh_msg_ctx *ctx,
			   struct net_buf_simple *buf);
```
we can assign only this type of function pointer. So i think we can't implement
your suggested way.

vikrant8051 replied:
Keep old implementation ....that means separate message handler for ack &unack.
So done !!

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -437,123 +356,44 @@ static void gen_level_set_unack(struct bt_mesh_model *model,
+		if (model->op->opcode == BT_MESH_MODEL_OP_GEN_LEVEL_SET) {
+			gen_level_get(model, ctx, buf);
+		}
 		gen_level_publish(model);
```
Yes, you are right. But as of now I don't remember whether PTS test expecting it or not.

vikrant8051 replied:
I will removed it.

vikrant8051 replied:
done !!

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -2335,18 +1702,18 @@ void light_ctl_temp_publish(struct bt_mesh_model *model)
-	if (state->temp == state->target_temp &&
-	    state->delta_uv == state->target_delta_uv) {
+	if (state->temp->current == state->temp->target &&
+	    state->duv->current == state->duv->target) {
```
Yes, in case of all _publish it is redundant to compare current & target value.
I will replace it by

if (state->transition->counter == 0) {
    goto publish;
} else {
}

But same we can't do in case of all _get()

vikrant8051 replied:
Yes. it is redundant. We should call publish only when target /= current. In that case, it is not necessary to compare them. I will remove it.

vikrant8051 replied:
done !!

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c:
```diff
@@ -64,43 +61,40 @@ static u16_t linear_to_actual(u16_t val)
+	    ctl->light->target < ctl->light->range_min) {
+		ctl->light->target = ctl->light->range_min;
+	} else if (ctl->light->target > ctl->light->range_max) {
+		ctl->light->target = ctl->light->range_max;
```
ok

vikrant8051 replied:
done !!

vikrant8052 committed:
samples: mesh: nrf52: removed redundant coding

Removed redundant coding which was related to old implementation.
Now Server will publish the new state information to the model’s Publish
Address only when there is mismatch between target & current values.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
sample: mesh: nrf52: coding style improvement

Adjusted lines which are crssing limit of 80 characters to remove
check Warnings.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -437,123 +356,44 @@ static void gen_level_set_unack(struct bt_mesh_model *model,
-		/* Secondary element */
-		state->transition->type = LEVEL_TEMP_TT;
-		level_temp_handler(state);
-	}
```
As of now, I've removed **model->op->opcode** based implementation.

vikrant8051 replied:
done !!

vikrant8052 committed:
samples: mesh: nrf52: corrected  Gen. move msg handler implementation

Corrected Generic Move (Level) message handler implementation.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: removed unnecessary global vaiable

Removed unnecessary global variable & replaced them with different
variable defined in struct light_ctl_state.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
Thank you @trond-snekvik for your suggestions (it really helped to remove the bugs & lots of redundancy in overall implementation which I couldn't figured out due to code blindness) & review.
